### PR TITLE
MINIFICPP-67: Merge Content processor

### PR DESCRIPTION
**Pull request by minifirocks**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by benqiu2016** [comment]
Please review and provide feedback.

**Comment by phrocker** [comment]
@benqiu2016 I"m currently reviewing, but have yet to finish it, at which point the comments will open up.

File: libminifi/include/processors/BinFiles.h [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+// Bin Class
+class Bin {
+ public:
+  // Constructor
+   /*!
+    * Create a new Bin
```
[the remark this answers is not in the document]
Comment (benqiu2016) [reply libminifi/include/processors/BinFiles.h]: will do

File: libminifi/include/processors/BinFiles.h [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+  }
+  // check whether the bin is full
+  bool isFull() {
+    if (queued_data_size_ >= maxSize_ || queue_.size() >= maxEntries_)
+      return true;
+    else
+      return false;
+  }
+  // check whether the bin meet the min required size and entries
+  bool isFullEnough() {
```
[the remark this answers is not in the document]
Comment (benqiu2016) [reply libminifi/include/processors/BinFiles.h]: i minic with java did, it indicated whether the bin's full enough (meet the min requirement) for processing.

File: libminifi/include/processors/BinFiles.h [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+namespace processors {
+
+// Bin Class
+class Bin {
+ public:
+  // Constructor
+   /*!
+    * Create a new Bin
+    */
+  explicit Bin(uint64_t minSize, uint64_t maxSize, int minEntries, int maxEntries, std::string fileCount, std::string groupId)
```
[the remark this answers is not in the document]
Comment (benqiu2016) [reply libminifi/include/processors/BinFiles.h]: sure, will do.

File: libminifi/include/processors/BinFiles.h [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+      return true;
+    else
+      return false;
+  }
+  // check whether the bin meet the min required size and entries
+  bool isFullEnough() {
+    return isFull() || (queued_data_size_ >= minSize_ && queue_.size() >= minEntries_);
+  }
+  // check whether the bin is older than the time specified in msec
+  bool isOlderThan(uint64_t duration) {
```
[the remark this answers is not in the document]
Comment (benqiu2016) [reply libminifi/include/processors/BinFiles.h]: sure.

File: libminifi/include/processors/BinFiles.h [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+  std::string fileCount_;
+  std::string groupId_;
+  std::shared_ptr<logging::Logger> logger_;
+  // A global unique identifier
+  uuid_t uuid_;
+  // UUID string
+  std::string uuid_str_;
+};
+
+// BinManager Class
```
[the remark this answers is not in the document]
Comment (benqiu2016) [reply libminifi/include/processors/BinFiles.h]: Bin Manager has control of the access cocurrency of the bin file.
bin file is not thread safe. BinManager is thread safe.

File: libminifi/include/processors/BinFiles.h [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+  // Constructor
+   /*!
+    * Create a new BinManager
+    */
+  explicit BinManager()
+      : minSize_(0), maxSize_(ULLONG_MAX), maxEntries_(INT_MAX), minEntries_(1), binAge_(ULLONG_MAX), binCount_(0),
+        logger_(logging::LoggerFactory<BinManager>::getLogger()) {
+  }
+  virtual ~BinManager() {
+    purge();
```
[the remark this answers is not in the document]
Comment (benqiu2016) [reply libminifi/include/processors/BinFiles.h]: OK.

File: libminifi/include/processors/BinFiles.h [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+  }
```
[the remark this answers is not in the document]
Comment (benqiu2016) [reply libminifi/include/processors/BinFiles.h]: ok.

File: libminifi/include/processors/BinFiles.h [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+  std::string fileCount_;
+  // Bin Age in msec
+  uint64_t binAge_;
+  std::map<std::string, std::unique_ptr<std::deque<std::unique_ptr<Bin>>>> groupBinMap_;
+  std::deque<std::unique_ptr<Bin>> readyBin_;
+  int binCount_;
+  std::shared_ptr<logging::Logger> logger_;
+};
+
+// BinFiles Class
```
[the remark this answers is not in the document]
Comment (benqiu2016) [reply libminifi/include/processors/BinFiles.h]: i want to put into the single header file such that it is easy for user to locate the same.

**Comment by minifirocks** [comment]
@apiri please review the same while Marc is out. I addressed his comments

**Comment by minifirocks** [comment]
@phrocker @apiri please approve the same, i would like to merge that before i add the compression/tar support for merge content.

**Comment by phrocker** [comment]
@minifirocks Can you rebase against master? Your commits have unrelated content.

**Comment by minifirocks** [comment]
@phrocker rebased

**Comment by minifirocks** [comment]
@phrocker @apiri please let me know whether it can be merged to master.

**Comment by phrocker** [comment]
@minifirocks I did have a question that popped up after I hit approve....with the header and footer you're serializing the data directly. Did you have any issues when opening those merged content files in NiFi after it was sent via Site To Site?

**Comment by minifirocks** [comment]
@phrocker i did not send the file over site2site, i save to a file and i can open it for read OK. the sending of these flowfile is the same procedure that we use to send normal flow files.

**Comment by phrocker** [comment]
@minifirocks What did you use to open it?  The procedure is the same, but the interpretation of that data is not. Seems that we need to at least verify that NiFi can interpret what we send it.

**Comment by minifirocks** [comment]
@phrocker i tie the merge processor to a put file processor to save the content

**Comment by minifirocks** [comment]
@phrocker let me test the NiFi site2site to double check

**Comment by minifirocks** [comment]
@phrocker test the site2site, NiFi receive the right header/footer/demarcator

**Comment by minifirocks** [comment]
@phrocker @apiri please take a look and see whether you can merge the PR to master. i tested the site2site as above.

**Comment by phrocker** [comment]
@minifirocks thanks! I'll merge to master at some point today

**Comment by minifirocks** [comment]
@phrocker thanks for the help

**Pull request closed without merging**